### PR TITLE
Add type hints for generators and accessors

### PR DESCRIPTION
**Pull request by maxrjones**
**Description of proposed changes**

Making it easier to catch errors, for example for dataset/dataarray differences.

Fixes #

### CODE REVIEW

File: xbatcher/generators.py [hunk 1]
```diff
@@ -167,22 +178,20 @@ def _gen_batches(self) -> dict:
...
                 batches.append(list(input_generator))
             else:
                 batches += list(input_generator)
 
-        return dict(zip(range(len(batches)), batches))
+        return dict(enumerate(batches))
 
-    def _iterate_batch_dims(self, ds):
-        return _iterate_through_dataset(ds, self.batch_dims)
+    def _iterate_batch_dims(self) -> Any:
```
Comment (jhamman) [inline xbatcher/generators.py:193]:
```suggestion
    def _iterate_batch_dims(self) -> Generator[Dict[str, int], None, None]:
```

I didn't test this but I think it is basically right. If it works, you can propagate this to the `_iterate_through_dataset`  function as well.

Comment (jhamman) [reply xbatcher/generators.py:193]: Alternatively, you can use the `Iterator` type.

**Comment by jhamman** [comment]
@maxrjones - this is great. Is mypy active in one of the CI environments? I didn't see it but I could have missed it.

**Comment by maxrjones** [comment]
> @maxrjones - this is great. Is mypy active in one of the CI environments? I didn't see it but I could have missed it.

Thanks for the reviews @jhamman! Mypy is activate via the pre-commit CI (e.g., https://results.pre-commit.ci/run/github/155272397/1668812992.ah7CK-WJSvCUCBL9ANdLhA).

File: xbatcher/generators.py (hunk 1 above)
Comment (maxrjones) [reply xbatcher/generators.py:193]: Lots of errors using `def _iterate_batch_dims(self) -> Iterator[Dict[Hashable, slice]]:`.
```
xbatcher/generators.py:182: error: Keywords must be strings
xbatcher/generators.py:182: error: Argument 1 to "isel" of "Dataset" has incompatible type "**Dict[Hashable, slice]"; expected "Optional[Mapping[Any, Any]]"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "Dataset" has incompatible type "**Dict[Hashable, slice]"; expected "bool"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "Dataset" has incompatible type "**Dict[Hashable, slice]"; expected "Literal['raise', 'warn', 'ignore']"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "DataArray" has incompatible type "**Dict[Hashable, slice]"; expected "Optional[Mapping[Any, Any]]"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "DataArray" has incompatible type "**Dict[Hashable, slice]"; expected "bool"
xbatcher/generators.py:182: error: Argument 1 to "isel" of "DataArray" has incompatible type "**Dict[Hashable, slice]"; expected "Literal['raise', 'warn', 'ignore']"
xbatcher/generators.py:189: error: Argument 1 to "list" has incompatible type "Iterator[Dict[Hashable, slice]]"; expected "Iterable[List[Dict[Hashable, slice]]]"
Found 8 errors in 1 file (checked 20 source files)
```
I have a separate refactor almost finished that would fix https://github.com/xarray-contrib/xbatcher/issues/131 and make this function obsolete. Do view it as necessary to type all functions before merging in the existing type hints? My preference would be to not spend more time on this.

**Review by jhamman — APPROVED** [review]

**Pull request merged**